### DISCUSSION TITLE
Prometheus metrics

### DISCUSSION THREAD

**Discussion by lucasvbeek** [PurpurMC/Purpur#527] [__Dismissed__] [answered]
I think it'd be pretty cool if purpur had built-in support for exposing metrics, preferably in the Prometheus format.

Prometheus is a time-series database and monitoring solution, which can be coupled with grafana for data visualisation.

This would allow people to achieve the following:
 - Have (historical) insight to metrics like playercount, tps, cpu usage, ram usage etc
 - Properly monitor their server(s) and perhaps even set-up alerting, so they get notified when their tps drops for example

Technically it's just a http endpoint with data in a specific format, so people could even scrape the metrics on their own and create their own tools.

More information about the format and how prometheus works can be found here:
 - https://prometheus.io/docs/introduction/overview/
 - https://prometheus.io/docs/instrumenting/exposition_formats/
 - https://prometheus.io/docs/instrumenting/writing_exporters/

**Comment by lucasvbeek** [comment]
While I came to think of it, hosts could also make use of this and show metrics in their panel, and possibly even offer a basic monitoring service. That would be pretty awesome imo.

**Comment by katekerllenevich** [comment]
Is this not good enough for you? https://github.com/Cubxity/UnifiedMetrics

**Reply by lucasvbeek** [reply]
It's a pretty good plugin, but I think implementing a first-party solution would be better because it would:
1. reach more people, as it's a relatively unknown plugin, so more people can take advantage of it
2. be easier for hosts to implement, as the purpur user base is pretty significant so they don't have to support multiple smaller projects

**Comment by wouterdedroog** [comment]
Cool! Would love to see this feature added to Purpur.

**Comment by katekerllenevich** [comment, marked as answer]
I'm making an executive descision that this is out of scope for Purpur and it's better off in a plugin.
